anndata/_core/merge.py

```diff
@@ -141,7 +141,21 @@ def equal_dask_array(a, b) -> bool:
 
 @equal.register(np.ndarray)
 def equal_array(a, b) -> bool:
-    return equal(pd.DataFrame(a), pd.DataFrame(asarray(b)))
+    def _is_numeric_or_bool(x) -> bool:
+        """array_equal with equal_nan=True only works for these types."""
+        return np.issubdtype(x, np.number) or np.issubdtype(x, np.bool_)
+
+    b = asarray(b)
+
+    if a.shape != b.shape:
+        return False
+
+    if _is_numeric_or_bool(a.dtype) and _is_numeric_or_bool(b.dtype):
+        return np.array_equal(a, b, equal_nan=True)
+    else:
+        # For non numeric types use pandas to compare
+        # Reshapeing allows us to compare inputs with >2 dimensions
+        return equal(pd.DataFrame(a.reshape(-1)), pd.DataFrame(b.reshape(-1)))
```

Comment:
```suggestion
    if _is_numeric_or_bool(a.dtype) and _is_numeric_or_bool(b.dtype):
        return np.array_equal(a, b, equal_nan=True)
    # For non numeric types use pandas to compare
    # Reshapeing allows us to compare inputs with >2 dimensions
    return equal(pd.DataFrame(a.reshape(-1)), pd.DataFrame(b.reshape(-1)))
```

`else` is unnecessary here.

Reply:
I think this is more of a style thing. I think having the returns under their own conditions is a little more clear that the function could have returned higher up.

And broadly if the linter says it's fine, I think it's fine.